src/frontend/src/common/utils/locationUtils.ts

```diff
@@ -151,3 +151,70 @@ export const getIdFromCollectionLocation = (
     return collectionLocation.id;
   }
 };
+
+function findMaxDepthLocation(
+  searchLocation: GisaidLocation | NamedGisaidLocation,
+  maxDepth: keyof GisaidLocation,
+  locations: NamedGisaidLocation[],
+  locationFinderCache: LocationFinderCache | null = null
+): NamedGisaidLocation | undefined {
+  // Safety check -- Should not occur, but if app hasn't finished loading the
+  // Locations data and this gets called, would blow up because we assume there
+  // are some Locations below. Instead, fallback to `undefined` location.
+  if (locations.length === 0) {
+    return undefined;
+  }
+
+  // Only do cache interactions if one was provided
+  if (
+    locationFinderCache !== null &&
+    locationFinderCache[`${searchLocation.id}${maxDepth}`]
+  ) {
+    return locationFinderCache[`${searchLocation.id}${maxDepth}`];
+  }
+
+  const locationHierarchy: (keyof GisaidLocation)[] = [
+    "region",
+    "country",
+    "division",
+    "location",
+  ];
+  const finalIndex = locationHierarchy.findIndex(
+    (element) => element === maxDepth
+  );
+
+  const foundLocations = locationHierarchy.reduce((acc, tier, index) => {
+    // Until we get to the index of the maxDepth, search for the searchLocation value
+    // After that, search for null
+    const filterValue = index <= finalIndex ? searchLocation[tier] : null;
+    return acc.filter((value) => value[tier] === filterValue);
+  }, locations);
```

Comment:
There is an optimization here where we don't need to check further tiers if the previous tier is null - i.e. if division is null, don't check location.  It's not coming to me though... also not sure how relevant this is since we're currently using max depth of "location" for sc2 and "division" for mpox.

Reply:
This is kind of orthogonal, but it feels like the easiest approach is to just ensure that whatever provides the locations upstream pre-filters the Locations list down to  already dealing with the `maxDepth` aspect, so nothing of greater depth is provided in the Locations list already. The big challenge there is how do we deal with the group's location, but maybe we could just put in a special look-up somehow? Although isn't that already happening via the BE by passing the query param?

I guess my above thought could screw things up for TSV import. If a user puts in "Los Angeles" for their location in TSV when doing Mpox, should "California" get populated (b/c Mpox max depth is division), or should just nothing get populated at all? How I'd approach this depends on how we need to handle metadata import regarding locations, but I just don't know what the specs are there.